File: api/app/core/money.py
"""Exact money. The Python mirror of `dif/src/money.rs`.

Internally money is **integer paise** (1 INR = 100 paise). Floats are never used for
money math. Conversions go through `Decimal` with explicit rounding so results are exact
and reproducible — a wrong paise is a compliance defect, not a rounding nicety.
"""

from __future__ import annotations

from decimal import ROUND_HALF_UP, Decimal

_HUNDRED = Decimal(100)


class Paise(int):
    """An amount in integer paise. ``Paise.from_rupees("150.50") == Paise(15050)``."""

    __slots__ = ()

    @classmethod
    def from_rupees(cls, rupees: str | int | float | Decimal) -> Paise:
        """Construct from a rupee amount. ``float`` is accepted but routed through
        ``Decimal(str(...))`` so e.g. ``0.1`` does not leak binary error."""
        d = Decimal(str(rupees)) * _HUNDRED
        return cls(int(d.quantize(Decimal("1"), rounding=ROUND_HALF_UP)))

    @classmethod
    def from_paise(cls, paise: int) -> Paise:
        return cls(int(paise))

    @property
    def rupees(self) -> Decimal:
        """Exact rupee value as ``Decimal`` — for display, never for further money math."""
        return Decimal(int(self)) / _HUNDRED

    def format_inr(self) -> str:
        """Indian-grouped rupee string, e.g. ``₹12,34,567.00`` (UI polish requirement)."""
        sign = "-" if self < 0 else ""
        whole, frac = divmod(abs(int(self)), 100)
        s = str(whole)
        if len(s) > 3:
            head, tail = s[:-3], s[-3:]
            # group the head in pairs (Indian system: ..,xx,xx,xxx)
            parts: list[str] = []
            while len(head) > 2:
                parts.insert(0, head[-2:])
                head = head[:-2]
            if head:
                parts.insert(0, head)
            s = ",".join(parts) + "," + tail
        return f"{sign}₹{s}.{frac:02d}"


ZERO = Paise(0)

File: api-nest/src/common/money.ts
/**
 * Exact money. TS mirror of api/app/core/money.py (which mirrors dif/src/money.rs).
 *
 * Money is integer paise (1 INR = 100 paise) end to end — never a float. JS numbers are
 * exact to 2^53 paise (≈ ₹90 trillion), far above any real figure. `moneyColumn` stores
 * paise as Postgres BIGINT but hands JS a `number` (not the string TypeORM defaults to),
 * so arithmetic stays numeric and exact.
 */
import { ColumnOptions } from 'typeorm';

const HUNDRED = 100;

/** Paise from a rupee amount, half-up (mirrors Decimal(str(x))*100 quantize ROUND_HALF_UP). */
export function paiseFromRupees(rupees: string | number): number {
  // Scale on the decimal string, not a binary float: "1.005" → 101 paise (float*100 would give 100).
  const s = (typeof rupees === 'string' ? rupees : rupees.toString()).trim();
  const neg = s.startsWith('-');
  const [intPart = '0', fracPart = ''] = s.replace(/^[-+]/, '').split('.');
  const base = Number(intPart || '0') * HUNDRED + Number((fracPart + '00').slice(0, 2) || '0');
  const roundUp = (fracPart[2] ?? '0') >= '5' ? 1 : 0; // ROUND_HALF_UP on the sub-paise digit
  const total = base + roundUp;
  return neg ? -total : total;
}

/** Indian-grouped rupee string, e.g. ₹12,34,567.00. Display only — never money-math input. */
export function formatInr(paise: number): string {
  const sign = paise < 0 ? '-' : '';
  const whole = Math.floor(Math.abs(paise) / 100);
  const frac = Math.abs(paise) % 100;
  let s = String(whole);
  if (s.length > 3) {
    let head = s.slice(0, -3);
    const tail = s.slice(-3);
    const parts: string[] = [];
    while (head.length > 2) {
      parts.unshift(head.slice(-2));
      head = head.slice(0, -2);
    }
    if (head) parts.unshift(head);
    s = parts.join(',') + ',' + tail;
  }
  return `${sign}₹${s}.${String(frac).padStart(2, '0')}`;
}

/** BIGINT paise column that round-trips as a JS number. Use for every money column. */
export function moneyColumn(opts: ColumnOptions = {}): ColumnOptions {
  return {
    type: 'bigint',
    transformer: {
      to: (v: number | null) => v,
      from: (v: string | null) => (v === null || v === undefined ? v : Number(v)),
    },
    ...opts,
  };
}
